### DISCUSSION TITLE
converting cue schema to go types

### DISCUSSION THREAD

**Discussion by yassinm** [cue-lang/cue#1889] [Q&A] [answered]
Hi All.

I am trying to convert a schema defined by a user as below using go.

```
  schema: {
    fVar: float
    strVar: string
  }
```

I am trying to make the schema definition as simple as possible and not force the user to add more than hey have to when editing the  cue file. Whats the best way to do this ? Each field in this struct is returning a `BottomKind` when i call kind on each field.

I looked at the code in `encoding/gocode/generator.go` but that assumes the go struct is available to make the conversion. Where as in my case i don't want to ask the user for the corresponding go struct.

tbc all i am trying to do is hot to get the type information from each field assuming i have a `cue.Value` ?

**Comment by verdverm** [comment, marked as answer]
You want to use the `value.IncompleteKind()` function rather than just the `value.Kind()`, there is a bit of a misnomer in the function names. The first (longer) name has a much wider set of types it returns.

Regarding the Go Codec, it only generates helper functions to validate existing types with a CUE based schema. The CUE team has intentions to expand the codec, with one such enhancement being generating the Go structs for a value or definition, tbd what this looks like. There are a lot of complications once you get into the details, and even more if you want to generate types in other languages.

I've written up some of the complications and my initial thoughts around it here: https://docs.hofstadter.io/code-generation/type-and-structs/ and working on a prototype here: https://github.com/hofstadter-io/hofmod-types The strategy `hof` uses is to have users write in a DSL, rather than CUE values with incomplete types. (i.e.  `name: "string"` vs `name: string`)

**Comment by yassinm** [comment]
@verdverm thank you so much for that detailed answer.

I am trying to see if i can complete what i am doing using incomplete cue values and not go for a DSL. I find the cue values are way more natural to write

From an API perspective it would be nice to have reflection as part of the go library.  I assume a bunch of people will be trying to use cue as the source of truth for schema/interchange definition. I am in that category and i am sure i have a ton of bias and could be wrong.

**Reply by verdverm** [reply]
Yes, I agree, some thoughts...

To generate types, you eventually have to write code in the target language.

This involves writing string content to a file. So how do you do this? One way is to write code which builds up the string, but this means to target new languages, you have to write more target language specific code in Go. This also would give you full access to CUE's Go API to do much more sophisticated things. However, people can write some pretty fancy CUE and trying to write code which can handle all of these cases is quite difficult if not impossible. I expect this path to improve with time.

The other main strategy is to use `text/template`. The problem here is that these templating systems only support concrete data, so you have to represent the CUE types differently. `hof` currently has users writing in a DSL directly, and provides some simple types you can import to use which will create a more complex value containing more information. I believe this path is far more flexible and simplifies the ability to build out an ecosystem. It is possible to transform the incomplete CUE values people want to write into the more complex DSL that the templates benefit from. There are many similar edge cases to the Go code path, and I also expect this path to improve as CUE matures. In particular the #943 (core builtin additions) will greatly help this effort.

**Reply by yassinm** [reply]
```
  schema: {
    iVar: int64
  }
  ```

This is currently being accepted by the cue parser `v0.4.3` but there is no way i can tell the difference between `int` and `int64` when i am converting the type. Is this something that can be worked around ?

I am thinking that maybe i will introduce a base cue library with all the types defined for each language and then the user can import that. at conversion time i can look for attributes that make the conversion easier. Does this make sense ? or their is a better way ?

```
import "cuelib/types/go"

  schema: {
    iVar: go.int64
  }

----
// cuelib/types/go

int64 :{
 @go(int64)
}

 ```

**Reply by verdverm** [reply]
I'm not sure, this is exactly the type of situation which causes the complexities and why I am going with DSL for the time being.

Using DSLs means I can maintain a stable interface as CUE changes and matures on the path to v1.

I was using attributes, but have since decided not to. I outlined why here: https://github.com/cue-lang/cue/discussions/1865#discussioncomment-3486721

Similar to your idea of having a package of types: https://github.com/hofstadter-io/hof/blob/_dev/schema/dm/fields/common.cue as the language agnostic version and https://github.com/hofstadter-io/hof/tree/_dev/schema/dm/enrichers for language specific mappings. One thing to think about (and it too is tricky) is higher level types like UUID or email, that users building apps think about, but languages don't handle themselves. These are pretty alpha and I have some changes already planned, notably how to deal with UUID and the necessary import (though the idea applies generally).
